### DISCUSSION TITLE
Using HASwitch object in an array

### DISCUSSION THREAD

**Discussion by michaelbouma** [dawidchyrzynski/arduino-home-assistant#38] [General]
Hi,

Im busy with creating some software for the arduino Mega 2560 and discovered this library. I have a lot of IO and wonder if it is possible to use for example HASwitch as an array? 

I'm using this

`HASwitch *myRelay[NumberOfRelays]; `

```
void relayInit()
{
  ShowDebug("Initialising Relays...");	
  for (int i = 0; i < NumberOfRelays ; i++) {
	  
	Serial.println(relays[i].pin,DEC);
	myRelay[i] = new HASwitch(relays[i].Name, true);

	pinMode(relays[i].pin, OUTPUT);
}

```
But I have trouble with using the callback. Is is possible to do it ithis way? And how to address the callback?

**Comment by dawidchyrzynski** [comment]
Hi @michaelbouma,

Yes, that's absolutely possible. I'm using the same approach in one of my projects. Just make sure that each switch is constructed before calling `mqtt.begin(...)` method. 

Regarding callback, you can define your logic inside the callback method and assign the same callback to all switches. The caller provides pointer to the HASwitch* as the second argument, so you can define your logic as following:
```cpp
void onSwitchStateChanged(bool state, HASwitch* s) {
    if (strcmp(s.uniqueId(), "uniqueId") == 0) { // uniqueId is the name you passed to the constructor
        // do something
    }
}
```

Please note that `HASwitch* s` pointer is the same one you assigned to `myRelay[]` array. You can also compare them like `if (myRelay[0] == s) { ... }`

**Reply by michaelbouma** [reply]
Thanks for the reply, works like a charm.

**Reply by pvoncken** [reply]
I tried do use this method on multiple FANs. 
Is that also possible?

**Comment by pvoncken** [comment]
Hi Michael,

I am building a multiple relais project and am very interested in your project. 
At the moment I can control 1 relay connected to my Mega Board from my Home Assistant Server.
But I need  to control 7 more Relay's but can't find out how to do it.
I have to admit that i'm prewtty new in coding arduino. 
Would you mind to share your Sketch?

**Reply by dawidchyrzynski** [reply]
Hi @pvoncken,

Maybe this will help you: https://github.com/dawidchyrzynski/arduino-home-assistant/issues/48

**Reply by pvoncken** [reply]
Works!
You're the best.
